apps/mark-scan/accessible-controller/src/controllerd.rs

```diff
@@ -0,0 +1,408 @@
+
+    let key: keyboard::Key;
+    match action {
+        Action::Pressed => match button {
+            Button::Select => {
+                key = keyboard::Key::Enter;
+            }
+            Button::Left => {
+                key = keyboard::Key::Left;
+            }
+            Button::Right => {
+                key = keyboard::Key::Right;
+            }
+            Button::Up => {
+                key = keyboard::Key::Up;
+            }
+            Button::Down => {
+                key = keyboard::Key::Down;
+            }
+            Button::Help => {
+                key = keyboard::Key::Q;
+            }
+            _ => {
+                return Err(CommandError::UnsupportedButton(button));
+            }
+        },
+        Action::Released => {
+            // Button release is a no-op since we already sent the keypress event
+            return Ok(());
+        }
+    }
+
+    send_key(device, key)
+}
+
+fn validate_connection(port: &mut Box<dyn SerialPort>) -> Result<(), io::Error> {
+    let echo_command = EchoCommand::new(vec![0x01, 0x02, 0x03, 0x04, 0x05]);
+    let echo_command: Vec<u8> = echo_command.into();
+    match port.write(&echo_command) {
+        Ok(_) => println!("Echo command sent"),
+        Err(error) => eprintln!("{error:?}"),
+    }
+
+    let mut serial_buf: Vec<u8> = vec![0; 20];
+
+    let start_time = Instant::now();
+    loop {
+        match port.read(serial_buf.as_mut_slice()) {
+            Ok(size) => {
+                let echo_response = &serial_buf[..size];
+                assert_eq!(
+                    echo_command, echo_response,
+                    "Received different response from echo command: {echo_response:x?}"
+                );
+
+                println!("Received valid echo command response");
+                return Ok(());
+            }
+            Err(ref e) if e.kind() == io::ErrorKind::TimedOut => (),
+            Err(e) => eprintln!("Error reading echo response: {e:?}"),
+        }
+
+        if start_time.elapsed() >= MAX_ECHO_RESPONSE_WAIT {
+            break;
+        }
+
+        thread::sleep(POLL_INTERVAL);
+    }
+
+    Err(io::Error::new(
+        io::ErrorKind::TimedOut,
+        "No echo response received",
+    ))
+}
+
+fn create_virtual_device() -> Device {
+    uinput::open(UINPUT_PATH)
+        .unwrap()
+        .name("Accessible Controller Daemon Virtual Device")
+        .unwrap()
+        .event(uinput::event::Keyboard::All)
+        .unwrap()
+        .create()
+        .unwrap()
+}
+
+fn main() {
+    // Open the serial port
+    let port = serialport::new(DEVICE_PATH, DEVICE_BAUD_RATE)
+        .timeout(POLL_INTERVAL)
+        .open();
+
+    println!("Opened controller serial port at {DEVICE_PATH}");
+
+    // Create virtual device for keypress events
+    let mut device = create_virtual_device();
+    println!("Created virtual device");
+
+    // Wait for virtual device to register
+    thread::sleep(Duration::from_secs(1));
```

Comment:
Is there a way to be sure the virtual device registered?

Reply:
Seems like it's possible with the underlying `uinput` C library but that functionality isn't ported to the `uinput` crate we use. It's also possible with just the filesystem, though that requires some parsing: https://stackoverflow.com/questions/34064047/how-to-get-name-path-of-uinput-created-device

I'll experiment with this a little when I'm in front of the hardware again.